pkg/internal/server/grpc/auth.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpc
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kauthorizer "k8s.io/apiserver/pkg/authorization/authorizer"
+
+	"github.com/kubernetes-csi/external-snapshot-metadata/pkg/internal/authn"
+	"github.com/kubernetes-csi/external-snapshot-metadata/pkg/internal/authz"
+)
+
+func (s *Server) authenticateAndAuthorize(ctx context.Context, token string, namespace string) error {
+	// Authenticate request with security token and find the user identity
+	authenticated, userInfo, err := s.authenticateRequest(ctx, token)
+	if err != nil {
+		return status.Errorf(codes.Internal, msgInternalFailedToAuthenticateFmt, err)
+	}
+	if !authenticated {
+		return status.Error(codes.Unauthenticated, msgUnauthenticatedUser)
+	}
+
+	// Authorize user
+	authorizer := authz.NewSARAuthorizer(s.kubeClient())
+	decision, reason, err := authorizer.Authorize(ctx, userInfo, namespace)
+	if err != nil {
+		return status.Errorf(codes.Internal, mgsInternalFailedToAuthorizeFmt, err)
+	}
+	if decision != kauthorizer.DecisionAllow {
+		return status.Errorf(codes.PermissionDenied, msgPermissionDeniedFmt, reason)
```

Comment:
FYI: I removed the erroneous use of `err.Error()` from the original code here - `err` is `nil` at this time!